### DISCUSSION TITLE
Deploy lambda in a VPC

### DISCUSSION THREAD

**Discussion by Mathdon** [artilleryio/artillery#2159] [Artillery]
Hi, is it possible to configure the lambda the gets deployed when running the command
artillery run-lambda \
   --region eu-west-1 \
   blitz.yml
   
To de deployed into a VPC that is in your AWS account?

**Comment by bernardobridge** [comment]
Hi @Mathdon,

Yes, you can run your test with these additional CLI flags:

- `--subnet-ids=id1,id2` for a comma-separated list of subnet ids
- `--security-group-ids=id1,id2` for a comma-separated list of security group ids.

This will then define the `VpcConfig` of the Lambda (same as the AWS Node SDK would use - basically this: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-vpcconfig.html).

**Reply by Mathdon** [reply]
Hi @bernardobridge 
Thanks for the info.
I have got it running in a lambda now, and tried this profile
  phases:
    - duration: 300
      arrivalRate: 50
      rampTo: 300
      
I had some what looked to be artillery errors:

Summary report @ 08:27:21(+0100)
--------------------------------

errors.EAI_AGAIN: .............................................................. 1
errors.EINVAL: ................................................................. 2
errors.EMFILE: ................................................................. 2121
http.codes.200: ................................................................ 28577

Do you know how many rps it artillery should be able to make?

I was getting similar errors when running locally and was hoping to be able to run more rps in a lambda

**Reply by bernardobridge** [reply]
Hi @Mathdon,

Glad it's running now 🚀 .

---

For your other issue, what we typically recommend is to spread the scenario over more workers - that's why we recommend using a distributed platform like Lambda or Fargate. For example, you could try the following phase config:

```yaml
phases:
  - duration: 300
    arrivalRate: 5
    rampTo: 30
```

over 10 workers, by running the CLI command with the flag `--count 10`. Artillery CLI will aggregate all the metrics for you in the machine where you trigger the test from.

_Note: 10 workers might be overkill here, I just used that as an example as the numbers were simple to show what I meant._

**Reply by Mathdon** [reply]
Hi @bernardobridge 
That worked a treat, thank you.
When I was running these tests locally I was using a custom script to sig4 sign the request to hit the API.
I see in the limitations for lambda:
Custom JS code provided with config.processor cannot yet use any third-party npm packages
I had this in my request
beforeRequest: 'addAmazonSignatureV4'

So I read this as that will no longer work when running in lambda.

The artillery-plugin-aws-sig4 plugin doesn't seem to work locally. do you know a way around this to be able to send sig4 signed requests from the lambda?


Really appreciate your help with this

**Reply by bernardobridge** [reply]
Hi @Mathdon. Awesome 💪!

---

Does your custom code depend on third-party npm packages? If so, then currently it wouldn't be able to run on Lambda, no. You'd have to write the code yourself using only native node packages or packages available in a Lambda environment (I think it comes with `aws-sdk` by default too).

Another alternative would be to run on `Fargate`, which should be just as straightforward to use, and doesn't have the same limitation as Lambda, so you can use third-party npm packages. You can also run Fargate tests in your own VPC (similar to Lambda).

**Reply by Mathdon** [reply]
Thanks for the reply, I'll try writing the code and see how I get on

thanks
